data/data-pipeline/data_pipeline/etl/base.py

```diff
@@ -116,63 +124,150 @@ def transform(self) -> None:
+                "Skipping validation step for this class because it does not "
+                "seem to be converted to new ETL class patterns."
+            )
+            return
 
-    def cleanup(self) -> None:
-        """Clears out any files stored in the TMP folder"""
+        if self.COLUMNS_TO_KEEP is None:
+            raise NotImplementedError(
+                "`self.COLUMNS_TO_KEEP` must be specified."
+            )
 
-        remove_all_from_dir(self.TMP_PATH)
+        if self.output_df is None:
+            raise NotImplementedError(
+                "The `transform` step must set `self.output_df`."
+            )
+
+        for column_to_keep in self.COLUMNS_TO_KEEP:
+            if column_to_keep not in self.output_df.columns:
+                raise ValueError(
+                    f"Missing column: `{column_to_keep}` is missing from "
+                    f"output"
+                )
+
+        for (
+            geo_level,
+            geo_field,
+            expected_geo_field_characters,
+            expected_rows,
+        ) in [
+            (
+                ValidGeoLevel.CENSUS_TRACT,
+                self.GEOID_TRACT_FIELD_NAME,
+                self.EXPECTED_CENSUS_TRACTS_CHARACTER_LENGTH,
+                self.EXPECTED_MAX_CENSUS_TRACTS,
+            ),
+            (
+                ValidGeoLevel.CENSUS_BLOCK_GROUP,
+                self.GEOID_FIELD_NAME,
+                self.EXPECTED_CENSUS_BLOCK_GROUPS_CHARACTER_LENGTH,
+                self.EXPECTED_MAX_CENSUS_BLOCK_GROUPS,
+            ),
+        ]:
+            if self.GEO_LEVEL is geo_level:
+                if geo_field not in self.COLUMNS_TO_KEEP:
+                    raise ValueError(
+                        f"Must have `{geo_field}` in columns if "
+                        f"specifying geo level as `{geo_level} "
+                    )
+                if self.output_df.shape[0] > expected_rows:
+                    raise ValueError(
+                        f"Too many rows: `{self.output_df.shape[0]}` rows in "
+                        f"output exceeds expectation of `{expected_rows}` "
+                        f"rows."
+                    )
+
+                geo_field_character_lengths = (
+                    self.output_df[geo_field].str.len().unique()
+                )
+
+                if len(geo_field_character_lengths) > 1:
+                    raise ValueError(
+                        f"Multiple character lengths for geo field "
+                        f"present: {geo_field_character_lengths}."
+                    )
+
+                elif (
+                    geo_field_character_lengths[0]
+                    != expected_geo_field_characters
+                ):
+                    raise ValueError(
+                        "Wrong character length: some of the census geography data "
+                        "has the wrong length."
+                    )
+
+                non_unique_geo_field_values = len(
```

Comment:
I think this would be clearer with `nunique` but it's a personal preference

```
self.output_df[geo_field].shape[0] == self.output_df[geo_field].nunique()
```